lib/resolve-value.js

```diff
@@ -41,66 +41,69 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
 
 	// Resolve any var(...) substitutons
 	var isResultantValueUndefined = false;
-	resultantValue = resultantValue.replace(new RegExp(RE_VAR_FUNC.source, 'g'), function(match, variableName, fallback) {
-		// Loop through the list of declarations for that value and find the one that best matches
-		// By best match, we mean, the variable actually applies. Criteria:
-		//		- is under the same scope
-		//		- The latest defined `!important` if any
-		var matchingVarDeclMapItem;
-		(map[variableName] || []).forEach(function(varDeclMapItem) {
-			// Make sure the variable declaration came from the right spot
-			// And if the current matching variable is already important, a new one to replace it has to be important
-			var isRoot = varDeclMapItem.parent.type === 'root' || varDeclMapItem.parent.selectors[0] === ':root';
 
-			var underScope = isNodeUnderScope(decl.parent, varDeclMapItem.parent);
-			var underScsopeIgnorePseudo = isNodeUnderScope(decl.parent, varDeclMapItem.parent, ignorePseudoScope);
-
-			//console.log(debugIndent, 'isNodeUnderScope', underScope, underScsopeIgnorePseudo, generateScopeList(varDeclMapItem.parent, true), varDeclMapItem.decl.value);
-
-			if(
-				underScsopeIgnorePseudo &&
-				// And if the currently matched declaration is `!important`, it will take another `!important` to override it
-				(!(matchingVarDeclMapItem || {}).isImportant || varDeclMapItem.isImportant)
-			) {
-				matchingVarDeclMapItem = varDeclMapItem;
+	while (resultantValue.match(new RegExp(RE_VAR_FUNC.source, 'g'))) {
```

Comment:
Can you fill me in on why the `resultantValue.replace(new RegExp(RE_VAR_FUNC.source, 'g')` is no longer suitable?

Reply:
the intention was to create a recursive match, to make sure to replace all variables correctly.

the previous version was not working with some nested css functions.

Reply:
The recursive stuff should come from the `resolveValue` calls inside 🤔 (see below)

Reply:
In my previous approaches, the original `RegExp` used was not covering this case, for example:

Original

```css
.box-foo {
	width: calc(58.3333333333% - var(--some-width, 100px)); 
}
```

Result: 

```css
.box-foo {
width: calc(58.3333333333% - var(--some-width, 100px); 
}
```

---

I changed the previous regex to `var RE_VAR_FUNC = (/var\(\s*(--[^,\s]+?)(?:\s*,\s*([^()]+|.+))?\s*\)/);
`, it solved the case above, but it wasn't working now for this case:

```
.box-foo {
	width: calc(58.3333333333% - var(--missing, var(--some-width, 100px))); 
}
```
Since the regex now was treating each `var()` reparately, I had to force recursive match.

So... I though to use a `while` block. Not sure if was a clever idea.

---

But I decided to change the regex, to match fill only the `variablesUsedInValue` array and use `match-recursive` lib to help me to match  recursively.  Right now I'm not sure if the `while` block is relevant or not.

It makes sense to you?

Reply:
It seems like the recursive bits are duplicated now,

Here is where the `var()` in the fallback should be resolved, https://github.com/MadLittleMods/postcss-css-variables/blob/fe5c821b8feedbe80d8f4399a2b80ca166bb3515/lib/resolve-value.js#L71-L76

And here is where we recursively resolve `var()` that resolve to more `var()` https://github.com/MadLittleMods/postcss-css-variables/blob/fe5c821b8feedbe80d8f4399a2b80ca166bb3515/lib/resolve-value.js#L92

It's still unclear to me why that can't be adapted to cover this new case. We may need some sort of `balanced-match` sort of dependency for proper parenthesis pairs but the overall code structure seems like it doesn't need to change to a `while` (this is just masking the bug in the current code)

Reply:
Yeah, I understand your point about the while. It seems redundant in some cases.

---

I tried `balanced-match` before, but the lib does not have a recursive feature. The reason why I choose `match-recursive` lib is that the lib returns an array of what was found.

First, let's assume this particular case:

```css
:root {
  --some-width: 150px;
  --some-other-width: 10px
}
.box-foo {
  width: calc(var(--some-width) - var(--some-other-width)); 
}
```

Using `balanced-match`, the returned output is:
```css
.box-foo {
	width: calc(150px - var(--some-other-width));
}
```

Using `match-recursive`, the returned output is:
```css
.box-foo {
	width: calc(150px - 10px);
}
```

To make these test I used the following functions to match:

```js
var balanced = require('balaced-match');
var matchRecursive = require('match-recursive');

balanced(/\bvar\(/i, /\)/, resultantValue);

matchRecursive(resultantValue, 'var(...)'))

```


It seems `balanced-match` returns only the **first occurrence** of a `var()` declaration. Meanwhile, if I use `match-recursive` lib, I can replace the whole sentence and retrieve the expected value.  Which leads me to understand that `balanced-match` would be the best choice for simple cases of nested functions, whereas `match-recursive` is best for complex cases of `var()` use, with more than one `var()` in the same sentence.

Reply:
I am starting to understand the new `while` loop now. It's about matching `var()`, not resolving. Sorry it took me a while to get my head around it properly.

I'd be more comfortable using [`balanced-match`](https://www.npmjs.com/package/balanced-match) given it's ubiquity. We can also use it like we were before going one by one AFAICT

```js
var balanced = require('balanced-match');

// not tested
while (match = balanced('var(', ')', resultantValue)) {
```

Seems like we should also use this same looping strategy to populate `variablesUsedInValueMap` just above

Reply:
I made changes and explained a little in https://github.com/MadLittleMods/postcss-css-variables/pull/97#issuecomment-487770935

Could you check?